subprojects/android-test/test-inhouse-runner/src/main/kotlin/com/avito/android/runner/annotation/validation/TestMetadataValidator.kt

```diff
@@ -0,0 +1,36 @@
+package com.avito.android.runner.annotation.validation
+
+import android.os.Bundle
+import com.avito.android.runner.annotation.resolver.MethodStringRepresentation
+import com.avito.android.runner.annotation.resolver.TestKindExtractor
+import com.avito.android.runner.annotation.resolver.getClassOrThrow
+import com.avito.report.model.Kind
+
+internal class TestMetadataValidator {
+
+    fun validate(instrumentationArguments: Bundle) {
+        val test = instrumentationArguments.getString("class")
+
+        if (test.isNullOrBlank()) {
+            throw RuntimeException("Test name not found in instrumentation arguments: $instrumentationArguments")
+        }
+        val parsedTest = MethodStringRepresentation.parseString(test)
+
+        @Suppress("NON_EXHAUSTIVE_WHEN")
+        when (TestKindExtractor.extract(parsedTest)) {
+            Kind.UI_COMPONENT, Kind.UI_COMPONENT_STUB, Kind.E2E_STUB -> validateNoSideEffects(test)
+        }
+    }
+
+    private fun validateNoSideEffects(test: String) {
+        val testClass = MethodStringRepresentation.parseString(test).getClassOrThrow()
+
+        val rules = NoSideEffectsRulesValidator.validate(testClass)
+        if (rules.isNotEmpty()) {
```

Comment:
Did you check how a user will see that error if he/she runs tests from `Android Studio`?

Reply:
Users can see such errors only in logcat. This is a general restriction. AS can't show errors from `AndroidJUnitRunner` and test application. 
I.e. I know no options on how to pass an error for a test to IDE.